app/src/main/java/org/mozilla/focus/fragment/HomeFragment.kt

```diff
@@ -172,19 +175,35 @@ private class HomeTileAdapter(
 private fun onBindBundledHomeTile(holder: TileViewHolder, tile: BundledHomeTile) = with (holder) {
     val bitmap = BundledTilesManager.getInstance(itemView.context).loadImageFromPath(itemView.context, tile.imagePath)
     iconView.setImageBitmap(bitmap)
+
+    titleView.text = tile.title
 }
 
 private fun onBindCustomHomeTile(uiLifecycleCancelJob: Job, holder: TileViewHolder, item: CustomHomeTile) = with (holder) {
-    launch(uiLifecycleCancelJob + CommonPool) {
-        val screenshot = HomeTileScreenshotStore.read(itemView.context, item.id) // TODO: if null, provide placeholder.
-        launch(UI) {
-            // Animate to avoid pop-in due to thread hand-offs. TODO: animation is janky.
-            ObjectAnimator.ofInt(iconView, "imageAlpha", 0, 255).apply {
-                interpolator = CUSTOM_TILE_ICON_INTERPOLATOR
-                duration = CUSTOM_TILE_ICON_TO_SHOW_MILLIS
-            }.start()
-            iconView.setImageBitmap(screenshot)
+    launch(uiLifecycleCancelJob + UI, CoroutineStart.UNDISPATCHED) {
+        val validUri = item.url.toJavaURI()
+
+        val screenshot = async { HomeTileScreenshotStore.read(itemView.context, item.id) }
+        val title = if (validUri == null) {
+            CompletableDeferred(item.url)
+        } else {
+            async { FormattedDomain.format(itemView.context, validUri, true, 1) }
         }
+
+        // Wait for both to complete so we can animate them together.
+        iconView.setImageBitmap(screenshot.await()) // TODO: if null, provide placeholder.
+        titleView.text = title.await()
+
+        // Animate to avoid pop-in due to thread hand-offs. TODO: animation is janky.
+        AnimatorSet().apply {
+            interpolator = CUSTOM_TILE_ICON_INTERPOLATOR
+            duration = CUSTOM_TILE_TO_SHOW_MILLIS
+
+            val iconAnim = ObjectAnimator.ofInt(iconView, "imageAlpha", 0, 255)
+            val titleAnim = ObjectAnimator.ofFloat(titleView, "alpha", 0f, 1f)
+
+            playTogether(iconAnim, titleAnim)
+        }.start()
```

Comment:
Do you need a timeout for this, or is it okay because either these files exist or they don't?

Reply:
For screenshots, we return if the file doesn't exist so no timeout necessary.

For public suffix, we crash if we're missing the file so no timeout necessary.

In theory, if this did get blocked, it wouldn't take up CPU (since it's suspended) and it'd get cancelled when the parent lifecycle job is cancelled anyway so it'll clean itself up.

Good question though. :)